### PR TITLE
feat: DataProcessor v1

### PR DESCRIPTION
**Pull request by dushyantbehl**
### Description of the change

This PR is to be merged after https://github.com/foundation-model-stack/fms-hf-tuning/pull/398 as it follows the changes. Once https://github.com/foundation-model-stack/fms-hf-tuning/pull/398 is merged this PR will be rebased on top of it.

This PR is to provide a data preprocessor framework which will enable flexibility to easily add more data preprocessing features in the future. This PR covers the following:

- Change base framework to the new configurable framework
   - Preprocessing utils function changed from `format_dataset` to `process_dataargs`
- Handle currently supported features using the new base framework
- Cleanup redundant code and port unit tests.
- Unit testing data preprocessor
- Unit testing setup data preprocessor

This PR does not explicitly enable any new features in fms-hf-tuning, it's purpose is to more easily allow new features to be added in the future by changing the backend.

Co-Authored by @willmj @Abhishek-TAMU 

### How to verify the PR

- New/refactored unit tests located in `testing/data`
- Unit tests of `train()` function
- Run e2e tuning + vLLM (see example [build](https://v3.travis.ibm.com/github/ai-foundation/fmaas-tuning-docs/builds/30211892))

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

File: tuning/sft_trainer.py [hunk 1]
```diff
@@ -317,7 +318,9 @@ def train(
         formatted_train_dataset,
         formatted_validation_dataset,
         dataset_text_field,
-    ) = format_dataset(data_args, tokenizer, max_seq_length)
+        #    ) = format_dataset(data_args, tokenizer, max_seq_length)
+    ) = process_dataargs(data_args, tokenizer, max_seq_length)
```
Comment (fabianlim) [inline tuning/sft_trainer.py:322]: this seems to be the only entrypoint into sft_trainer. But I cant understand how something like this can switch out the dataloader.

Comment (dushyantbehl) [reply tuning/sft_trainer.py:322]: @fabianlim fair point but this is because the patch is not complete yet..its `WIP` and I will add additional commandline args to take a data config like the one in ADR from the user. 
The patch was made public following our internal discussion to have @willmj  and @Abhishek-TAMU visibility into the implementation and helping with it if needed.

File: tuning/data/data_handlers.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+
+def apply_dataset_formatting(
+    element: Dict[str, str], tokenizer: AutoTokenizer, dataset_text_field: str, **kwargs
+):
+    return {
+        f"{dataset_text_field}": element[f"{dataset_text_field}"] + tokenizer.eos_token
+    }
+
```
Comment (Abhishek-TAMU) [inline tuning/data/data_handlers.py:62]: When [raw_datasets = raw_datasets.map(handler, **kwargs)](https://github.com/foundation-model-stack/fms-hf-tuning/blob/5e948f16a2c7fa4d942af434917d5483538827a0/tuning/data/data_processors.py#L166) is called and `kwargs["batched"] = True` then `element[f"{dataset_text_field}"]` here would be a list right. So does before condition make sense:

```
if "batched" in kwargs and kwargs["batched"]:
      return {
          f"{dataset_text_field}": element[f"{dataset_text_field}"] + tokenizer.eos_token
      }
  return {
      f"{dataset_text_field}": [
          text + tokenizer.eos_token for text in element[f"{dataset_text_field}"]
      ]
  }
```

File: tuning/utils/utils.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    _, ext = os.path.splitext(file_path)
+    return ext.lower()
+
+
+def get_loader_for_filepath(file_path: str) -> str:
+    ext = get_extension(file_path)
+    if ext == ".txt" | ".md":
+        return "text"
+    if ext == ".json" | ".jsonl":
+        return "json"
```
Comment (Abhishek-TAMU) [inline tuning/utils/utils.py:33]: Probably this I guess:

```
if ext in (".txt", ".md"):
    return "text"
if ext in (".json", ".jsonl"):
    return "json"
```

**Review by Abhishek-TAMU — COMMENTED** [review]
@dushyantbehl Thanks for the PR to let us know on the WIP. Just some comments on below would be appreciated.

File: tuning/data/data_handlers.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+# Third Party
+from transformers import AutoTokenizer
+
+# Local
+from tuning.utils.data_utils import custom_data_formatter
+from tuning.utils.preprocessing_utils import combine_sequence
+
+
+def tokenize_and_apply_instruction_masking(
```
Comment (Ssukriti) [inline tuning/data/data_handlers.py:28]:
```suggestion
def tokenize_and_apply_input_masking(
```

File: architecture_records/004-dataloader-v2.md
[the remark this answers is not in the document]
Comment (dushyantbehl) [reply architecture_records/004-dataloader-v2.md:1]: done

File: tuning/data/data_handlers.py (hunk 2 above)
Comment (dushyantbehl) [reply tuning/data/data_handlers.py:62]: @Abhishek-TAMU I consciously reverted this change and made our default handlers run in [`batched=False`](https://github.com/dushyantbehl/fms-hf-tuning/blob/96b94676038fe6206ef09ea192613f21a1a66b83/tuning/data/setup_dataprocessor.py#L94) mode for now, we can mention this in documentation for people to not use these handlers in `batched` mode. This change was made due to the fact that this and the other handlers we have defined right now are complex operations and need us to deconstruct each example from a batch before processing them. 

That being said this was after I had already implemented a patch to make all handlers take in `batched` and `non-batched` input and it is here in a [different branch](https://github.com/dushyantbehl/fms-hf-tuning/commit/ef2ef7808e547946bc395ce7cd83d1b71197c2b7)
so we can even cherry pick this to the current branch if needed. 

cc @Ssukriti @willmj @ashokponkumar

Comment (Abhishek-TAMU) [reply tuning/data/data_handlers.py:62]: > so we can even cherry pick this to the current branch if needed.

This sounds fine to me.

File: tuning/data/data_handlers.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    **tokenizer_kwargs,
+):
+    input = element[input_field_name]
+    output = element[output_field_name]
+
+    # TODO: Eventually move the code here
+    combined = combine_sequence(input, output, eos_token=tokenizer.eos_token)
+
+    tokenized_comb_seqs = tokenizer(combined, **tokenizer_kwargs)
+    tokenized_input = tokenizer(input, **tokenizer_kwargs)
```
[the remark this answers is not in the document]
Comment (Abhishek-TAMU) [reply tuning/data/data_handlers.py:42]: Made changes here: https://github.com/dushyantbehl/fms-hf-tuning/pull/2

**Comment by kmehant** [comment]
@dushyantbehl 
Support for packing pre-tokenized datasets is now first class feature in huggingface/trl. Good to allow for that in fms-hf-tuning.

fms-hf-tuning check previously present which can be relaxed and as well absorbed in to this PR : https://github.com/foundation-model-stack/fms-hf-tuning/blob/5c299408f6c2eea65f141709233f3ebc92d1b337/tuning/utils/preprocessing_utils.py#L85


PR for reference: https://github.com/huggingface/trl/pull/2011

File: tuning/data/setup_dataprocessor.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            "max_seq_length %s exceeds tokenizer.model_max_length \
+            %s, using tokenizer.model_max_length %s",
+            train_args.max_seq_length,
+            tokenizer.model_max_length,
+            tokenizer.model_max_length,
+        )
+
+    if data_args.data_config_path:
+        # Data config is specified so our processing path is diverging
+        return _process_dataconfig_file(
```
Comment (fabianlim) [inline tuning/data/setup_dataprocessor.py:123]: this is not very clear. what does this divergence mean? is this the legacy path?

File: tuning/data/setup_dataprocessor.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+            data_args, tokenizer, train_args.packing, max_seq_length
+        )
+
+    # Create a data processor with default loader config
+    default_loader_config = DataLoaderConfig()
```
Comment (fabianlim) [inline tuning/data/setup_dataprocessor.py:128]: I thought we are not calling this `DataLoader` anymore, because it conflates with HF's meaning of `DataLoader`?

File: tuning/data/setup_dataprocessor.py (hunk 6 above)
Comment (fabianlim) [reply tuning/data/setup_dataprocessor.py:123]: and its not clear to me if we updated the code to support non-json datasets, without resorting to the new data processor

File: tuning/data/data_preprocessing_utils.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    if not packing:
+        # TODO: near term - how response template ids are parsed out needs to be cleaned.
+        # The [2:] here applies if response template has \n prefix, it is needed to strip \n,
+        # otherwise template is not found. We will create issue to clean this out after we discuss
+        # data formats and collators we will support.
+        if response_template:
+            response_template_ids = tokenizer.encode(
+                response_template, add_special_tokens=False
+            )[2:]
+            return DataCollatorForCompletionOnlyLM(
```
Comment (fabianlim) [inline tuning/data/data_preprocessing_utils.py:174]: so we only support these 2 `collators`? will we support anymore in the other code paths? this will impact the padding free plugin

File: tuning/data/data_processors.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        self.registered_handlers[name] = func
+
+    @abstractmethod
+    def process_dataset_configs(
+        self, dataset_configs: List[DataSetConfig], **extra_kwargs
+    ) -> Union[Dataset, IterableDataset]:
+        raise NotImplementedError("Needs to be implemented")
+
+
+class HFBasedDataPreProcessor(DataPreProcessor):
```
Comment (fabianlim) [inline tuning/data/data_processors.py:67]: it is not clear, besides this one implemebntation, what other implementations we will need

File: tuning/data/data_preprocessing_utils.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        if is_traindata_tokenized:
+            return DataCollatorForSeq2Seq(
+                tokenizer=tokenizer, padding=True, max_length=max_seq_length
+            )
+        raise ValueError(
+            "Could not pick a data collator. Please refer to supported data formats"
+        )
+
+
+def custom_data_formatter(element, template, formatted_dataset_field):
```
Comment (fabianlim) [inline tuning/data/data_preprocessing_utils.py:190]: we need to document why this is needed? what not just use chat templates?

File: tuning/data/data_processors.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+
+
+def autoregister_available_handlers(processor: DataPreProcessor):
+    if processor is None:
+        return
+    for name, func in AVAILABLE_DATA_HANDLERS.items():
+        processor.register_data_handler(name=name, func=func)
+
+
+def get_dataprocessor(
```
Comment (fabianlim) [inline tuning/data/data_processors.py:226]: the function says `dataprocessro` but it returns a `dataPreProcessor` why the difference?

**Comment by fabianlim** [comment]
I feel the PR lacks docmentation. We should have some things written out for the common use cases
- adding a new data preprocessor
- adding a new handler. etc..
- fallback on for use cases that does not require the data preprocessor routines..
- how to write a data config.. etc.

File: tuning/data/setup_dataprocessor.py (hunk 6 above)
Comment (dushyantbehl) [reply tuning/data/setup_dataprocessor.py:123]: @fabianlim we had not initially planned large documentation for this feature because it was planned as a `dev preview` for this release while retaining the old behaviour (input output semantics) as we mentioned in the ADR. 

That being said, we will expand some comments in code to make the code path clear.

File: tuning/data/setup_dataprocessor.py (hunk 7 above)
Comment (dushyantbehl) [reply tuning/data/setup_dataprocessor.py]: Renamed. thanks for pointing this out.

File: tuning/data/data_preprocessing_utils.py (hunk 8 above)
Comment (dushyantbehl) [reply tuning/data/data_preprocessing_utils.py:174]: For now the data collators are not being touched by data preprocessor so this functionality is as it stands in the library currently. 
For future we will expand on collators as we grow use cases of data preprocessor.

File: tuning/data/data_processors.py (hunk 9 above)
Comment (dushyantbehl) [reply tuning/data/data_processors.py:69]: This one's a bit tricky because initially we wanted to keep the implementation modular to make sure this can be extended to any other requirements later (think maybe a data preprocessor built on top of custom torch datasets) but since we do not have any active requirements in the roadmap and such experimental features might not be directly integrated, I am dropping the abstract class in favour of just HF implementation.

File: tuning/data/data_preprocessing_utils.py (hunk 10 above)
Comment (dushyantbehl) [reply tuning/data/data_preprocessing_utils.py:190]: Again..this is not new functionality in the data preprocessor this is old code being retained. We can try to add some comments here to explain what the functions are doing slowly as we expand on the documetation.

**Comment by ashokponkumar** [comment]
> I feel the PR lacks docmentation. We should have some things written out for the common use cases
> 
> * adding a new data preprocessor
> * adding a new handler. etc..
> * fallback on for use cases that does not require the data preprocessor routines..
> * how to write a data config.. etc.

We definitely need it, but we will add it in the next PR where we will be exposing these features to external users. Currently these are internal implementation which is used by the existing exposed interface. 

So I would recommend going ahead with this PR and following it up with these documentations in the next PR.

File: tuning/data/setup_dataprocessor.py (hunk 6 above)
Comment (fabianlim) [reply tuning/data/setup_dataprocessor.py:123]: yes pls looking forward to it

File: tuning/data/data_preprocessing_utils.py (hunk 10 above)
Comment (fabianlim) [reply tuning/data/data_preprocessing_utils.py:190]: yes pls

**Review by fabianlim — COMMENTED** [review]
I am ok that my comments are addressed, cc: @ashokponkumar @Ssukriti

**Comment by dushyantbehl** [comment]
> @dushyantbehl Support for packing pre-tokenized datasets is now first class feature in huggingface/trl. Good to allow for that in fms-hf-tuning.
> 
> fms-hf-tuning check previously present which can be relaxed and as well absorbed in to this PR :
> 
> https://github.com/foundation-model-stack/fms-hf-tuning/blob/5c299408f6c2eea65f141709233f3ebc92d1b337/tuning/utils/preprocessing_utils.py#L85
> 
> PR for reference: [huggingface/trl#2011](https://github.com/huggingface/trl/pull/2011)
> 
> cc: @ashokponkumar

@kmehant this change has been reverted for now as we need to get `transformers` up to the version where this patch will be supported. Please open up an issue and we can take this up when `trl` is up to date.

**Review by willmj — APPROVED** [review]
LGTM
